File: sdk/python/captcha-background-sdk/captcha_background_sdk/background_index.py
from __future__ import annotations

from pathlib import Path
from typing import Dict, Iterable, List, Optional, Tuple

from PIL import Image

from .group_id import group_id_from_pixels
from .types import BackgroundMeta


_DEFAULT_IMAGE_EXTS = {".png", ".jpg", ".jpeg", ".webp", ".bmp"}


def load_rgba_pixels(image_path: str) -> Tuple[int, int, List[Tuple[int, ...]]]:
    with Image.open(image_path) as im:
        rgba = im.convert("RGBA")
        width, height = rgba.size
        return width, height, list(rgba.getdata())


def compute_group_id(image_path: str) -> str:
    width, height, pixels = load_rgba_pixels(image_path)
    return group_id_from_pixels(pixels, width, height)


def build_background_index(
    background_dir: str,
    recursive: bool = True,
    exts: Optional[Iterable[str]] = None,
) -> Dict[str, BackgroundMeta]:
    root = Path(background_dir)
    if not root.exists() or not root.is_dir():
        raise FileNotFoundError(f"background_dir not found: {background_dir}")

    valid_exts = _DEFAULT_IMAGE_EXTS if exts is None else {e.lower() for e in exts}
    files = root.rglob("*") if recursive else root.glob("*")

    index: Dict[str, BackgroundMeta] = {}
    for file_path in files:
        if not file_path.is_file() or file_path.suffix.lower() not in valid_exts:
            continue
        width, height, pixels = load_rgba_pixels(str(file_path))
        group_id = group_id_from_pixels(pixels, width, height)
        if group_id not in index:
            index[group_id] = BackgroundMeta(
                group_id=group_id,
                image_path=str(file_path),
                width=width,
                height=height,
            )
    return index


def build_background_index_from_files(
    image_paths: Iterable[str],
    exts: Optional[Iterable[str]] = None,
) -> Dict[str, BackgroundMeta]:
    valid_exts = _DEFAULT_IMAGE_EXTS if exts is None else {e.lower() for e in exts}
    index: Dict[str, BackgroundMeta] = {}
    for image_path in image_paths:
        file_path = Path(image_path)
        if not file_path.is_file() or file_path.suffix.lower() not in valid_exts:
            continue
        width, height, pixels = load_rgba_pixels(str(file_path))
        group_id = group_id_from_pixels(pixels, width, height)
        if group_id not in index:
            index[group_id] = BackgroundMeta(
                group_id=group_id,
                image_path=str(file_path),
                width=width,
                height=height,
            )
    return index

File: typescript-sdk/src/background-index.ts
import fs from 'fs';
import path from 'path';
import { BackgroundMeta } from './types';
import { groupIdFromPixels } from './group-id';
import { loadRgbaPixels } from './image-utils';

const DEFAULT_IMAGE_EXTS = new Set(['.png', '.jpg', '.jpeg', '.webp', '.bmp']);

export async function loadRgbaPixelsFromPath(imagePath: string) {
  return loadRgbaPixels(imagePath);
}

export async function computeGroupId(imagePath: string): Promise<string> {
  const { width, height, pixels } = await loadRgbaPixels(imagePath);
  return groupIdFromPixels(pixels, width, height);
}

async function scanFiles(root: string, recursive: boolean): Promise<string[]> {
  const entries = await fs.promises.readdir(root, { withFileTypes: true });
  const files: string[] = [];
  for (const entry of entries) {
    const fullPath = path.join(root, entry.name);
    if (entry.isDirectory()) {
      if (recursive) {
        files.push(...await scanFiles(fullPath, recursive));
      }
    } else if (entry.isFile()) {
      files.push(fullPath);
    }
  }
  return files;
}

export async function buildBackgroundIndex(
  backgroundDir: string,
  recursive: boolean = true,
  exts?: Iterable<string>
): Promise<Record<string, BackgroundMeta>> {
  if (!fs.existsSync(backgroundDir) || !fs.statSync(backgroundDir).isDirectory()) {
    throw new Error(`background_dir not found: ${backgroundDir}`);
  }
  const validExts = exts ? new Set(Array.from(exts).map((e) => e.toLowerCase())) : DEFAULT_IMAGE_EXTS;
  const files = await scanFiles(backgroundDir, recursive);

  const index: Record<string, BackgroundMeta> = {};
  for (const filePath of files) {
    const ext = path.extname(filePath).toLowerCase();
    if (!validExts.has(ext)) {
      continue;
    }
    const { width, height, pixels } = await loadRgbaPixels(filePath);
    const groupId = groupIdFromPixels(pixels, width, height);
    if (!index[groupId]) {
      index[groupId] = {
        group_id: groupId,
        image_path: filePath,
        width,
        height,
      };
    }
  }
  return index;
}

export async function buildBackgroundIndexFromFiles(
  imagePaths: Iterable<string>,
  exts?: Iterable<string>
): Promise<Record<string, BackgroundMeta>> {
  const validExts = exts ? new Set(Array.from(exts).map((e) => e.toLowerCase())) : DEFAULT_IMAGE_EXTS;
  const index: Record<string, BackgroundMeta> = {};
  for (const filePath of imagePaths) {
    if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
      continue;
    }
    const ext = path.extname(filePath).toLowerCase();
    if (!validExts.has(ext)) {
      continue;
    }
    const { width, height, pixels } = await loadRgbaPixels(filePath);
    const groupId = groupIdFromPixels(pixels, width, height);
    if (!index[groupId]) {
      index[groupId] = {
        group_id: groupId,
        image_path: filePath,
        width,
        height,
      };
    }
  }
  return index;
}
